src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -127,6 +128,216 @@ void CoreWorkerDirectTaskSubmitter::AddWorkerLeaseClient(
   RAY_CHECK(scheduling_key_entry.active_workers.size() >= 1);
 }
 
+void CoreWorkerDirectTaskSubmitter::ReturnWorker(const rpc::WorkerAddress addr,
+                                                 bool was_error,
+                                                 const SchedulingKey &scheduling_key) {
+  auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+  RAY_CHECK(scheduling_key_entry.active_workers.size() >= 1);
+  auto &lease_entry = worker_to_lease_entry_[addr];
+  RAY_CHECK(lease_entry.lease_client);
+  RAY_CHECK(lease_entry.tasks_in_flight == 0);
+  RAY_CHECK(lease_entry.stealable_tasks.size() == 0);
+
+  // Decrement the number of active workers consuming tasks from the queue associated
+  // with the current scheduling_key
+  scheduling_key_entry.active_workers.erase(addr);
+  if (scheduling_key_entry.CanDelete()) {
+    // We can safely remove the entry keyed by scheduling_key from the
+    // scheduling_key_entries_ hashmap.
+    scheduling_key_entries_.erase(scheduling_key);
+  }
+
+  struct timespec return_worker_time_;
+  clock_gettime(CLOCK_REALTIME, &return_worker_time_);
+  long double time_elapsed =
+      (long double)(return_worker_time_.tv_sec - initial_time_.tv_sec) +
+      (long double)((return_worker_time_.tv_nsec - initial_time_.tv_nsec) /
+                    (long double)1000000000.0);
```

Comment:
Agree! I added these logging instructions just so that I can check, after each revision, that the performance has not gone down. As you said, though, it wouldn't make sense to leave them in the code after we merge.